### PR TITLE
feat(object store): add list method

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

As titled.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/object_store/src/object/disk.rs [hunk 1]
```diff
@@ -249,6 +265,10 @@ impl ObjectStore for DiskObjectStore {
             .map_err(|e| ObjectError::disk(format!("failed to delete {}", path), e))?;
         Ok(())
     }
+
+    async fn list(&self, _prefix: &str) -> ObjectResult<Vec<ObjectMetadata>> {
+        unimplemented!()
```
Comment (wenym1) [inline src/object_store/src/object/disk.rs:270]: Why is it unimplemented? I think we can implement it through `listdir`.

File: src/object_store/src/object/mem.rs [hunk 2]
```diff
@@ -74,6 +78,10 @@ impl ObjectStore for InMemObjectStore {
         self.objects.lock().await.remove(path);
         Ok(())
     }
+
+    async fn list(&self, _prefix: &str) -> ObjectResult<Vec<ObjectMetadata>> {
+        unimplemented!()
```
Comment (wenym1) [inline src/object_store/src/object/mem.rs:83]: We can also implement it.

File: src/object_store/src/object/s3.rs [hunk 3]
```diff
@@ -114,6 +119,43 @@ impl ObjectStore for S3ObjectStore {
...
+    async fn list(&self, prefix: &str) -> ObjectResult<Vec<ObjectMetadata>> {
+        let mut ret: Vec<ObjectMetadata> = vec![];
+        let mut next_continuation_token = None;
+        loop {
+            let mut request = self
+                .client
+                .list_objects_v2()
+                .bucket(&self.bucket)
+                .prefix(prefix);
+            if let Some(continuation_token) = next_continuation_token.take() {
```
Comment (wenym1) [inline src/object_store/src/object/s3.rs:132]: What does this `continuation_token` mean? Could we have come comments explaining the general logic?

**Review by wenym1 — COMMENTED** [review]
Seems that the GC logic is included in this PR while adding `list` method is not depending on it. Could we move the GC logic to a new PR?

File: src/object_store/src/object/disk.rs (hunk 1 above)
Comment (zwang28) [reply src/object_store/src/object/disk.rs]: Fixed.

File: src/object_store/src/object/mem.rs (hunk 2 above)
Comment (zwang28) [reply src/object_store/src/object/mem.rs]: Fixed.

File: src/object_store/src/object/s3.rs (hunk 3 above)
Comment (zwang28) [reply src/object_store/src/object/s3.rs:135]: Fixed.

**Comment by zwang28** [comment]
> Seems that the GC logic is included in this PR while adding `list` method is not depending on it. Could we move the GC logic to a new PR?

Irrelevant code removed. PTAL.

**Review by wenym1 — APPROVED** [review]

**Pull request merged**